src/libraries/System.Numerics.Vectors/tests/Matrix4x4LeftHandChangedBasisTests.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Numerics.Tests
+{
+    public class Matrix4x4LeftHandChangedBasisTests : Matrix4x4RightHandTests
+    {
+        protected override string CreateLookAtMethodName => $"{nameof(Matrix4x4)}.{nameof(Matrix4x4.CreateLookAtLeftHanded)}";
+        protected override Matrix4x4 CreateLookAt(Vector3 cameraPosition, Vector3 cameraTarget, Vector3 cameraUpVector)
+        {
+            return LeftHandToRightHand * Matrix4x4.CreateLookAtLeftHanded(
+                ChangeRightHandToLeftHand(cameraPosition),
+                ChangeRightHandToLeftHand(cameraTarget),
+                ChangeRightHandToLeftHand(cameraUpVector)) * RightHandToLeftHand;
```

Comment:
`RightHandToLeftHand * CreateLookAtRightHanded(...)` gives a matrix that transforms from RH -> LH but `CreateLookAtLeftHanded(...)` gives a matrix that transforms from LH -> LH, so these won't be equal without another change of basis somewhere.

I could be misunderstanding though - are you suggesting to compare `RHtoLH * CreateLookAtRH * LHtoRH == CreateLookAtLH` instead of `RHtoLH * CreateLookAtRH * LHtoRH == expectedLookAtLH` (calling both Create methods instead of reusing the expected value)?

I think the classes in the PR are already structured like you described:
abstract class :> validate LH directly against expected output M<sub>L</sub> :> validate that using RH via transforms produces same expected output M<sub>L</sub>

abstract class :> validate RH directly against expected output M<sub>R</sub> :> validate that using LH via transforms produces same expected output M<sub>R</sub>

Reply:
Sorry, was conflating things a bit and it didn't come out clearly...

The general suggestion is that we should be checking the raw results for `Matrix4x4.*` and `Matrix4x4.*LeftHanded`. We should then also be checking that `Matrix4x4.*` transformed to a left-handed coordinate system is identical to `Matrix4x4.*LeftHanded` and vice versa.

Given that, we should be able to define a trivial `ChangeHandedness` helper that does what is "strictly correct" via `Inverse(P) * M * P` or which does a functional optimization based on knowledge that we're functionally just inverting `Z` such that it is equivalent in cost to a single matrix multiplication.

--------------------------------

To elaborate...

For this particular left vs right-handed coordinate system, we're defining that X and Y are identical and that Z is what is functionally inverted between them. Thus we're functionally using the following as the "change of basis" transform between the two:
```
 1  0  0  0
 0  1  0  0
 0  0 -1  0
 0  0  0  1
```

Changing the basis of the matrix is then strictly `Inverse(P) * M * P` where `P` is the change of basis and `M` is the existing matrix.

--------------------------------

Matrix multiplication `A * B` is then functionally (functionally transform each vector row `A.V` by matrix `B`):
```
A.X * B
A.Y * B
A.Z * B
A.W * B
```

Where `A.V * B` is then:
```
(A.V.X * B.X) + (A.V.Y * B.Y) + (A.V.Z * B.Z) + (A.V.W * B.W)
```
--------------------------------

Since the change of basis here is functionally an identity transform, just with `Z` negated we know that `Inverse(P)` is going to be functionally the same as `P` (minus some `-0` vs `+0` which aren't really relevant here). This is what allows us our first "cheat" and to essentially do `P * M * P` instead as we don't actually need to inverse `P` in this scenario. For other change of basis this may not strictly be true. For example, it is valid to define a change of basis that not only inverts the Z axis but which also scales everything by `2` in which case `Inverse(P) != P`.

Given the above and the simplification, however, the first step is:
```csharp
(P.X.X * M.X) + (P.X.Y * M.Y) + (P.X.Z * M.Z) + (P.X.W * M.W)
(P.Y.X * M.X) + (P.Y.Y * M.Y) + (P.Y.Z * M.Z) + (P.Y.W * M.W)
(P.Z.X * M.X) + (P.Z.Y * M.Y) + (P.Z.Z * M.Z) + (P.Z.W * M.W)
(P.W.X * M.X) + (P.W.Y * M.Y) + (P.W.Z * M.Z) + (P.W.W * M.W)
```

Filling in the known constants, we get:
```
( 1 * M.X) + ( 0 * M.Y) + ( 0 * M.Z) + ( 0 * M.W)
( 0 * M.X) + ( 1 * M.Y) + ( 0 * M.Z) + ( 0 * M.W)
( 0 * M.X) + ( 0 * M.Y) + (-1 * M.Z) + ( 0 * M.W)
( 0 * M.X) + ( 0 * M.Y) + ( 0 * M.Z) + ( 1 * M.W)
```

Which itself (ignoring NaN and -0 handling, which is irrelevant here) simplified down to:
```
 M.X
 M.Y
-M.Z
 M.W
```

The second part of the transform is then:
```
( M.X.X * P.X) + ( M.X.Y * P.Y) + ( M.X.Z * P.Z) + ( M.X.W * P.W)
( M.Y.X * P.X) + ( M.Y.Y * P.Y) + ( M.Y.Z * P.Z) + ( M.Y.W * P.W)
(-M.Z.X * P.X) + (-M.Z.Y * P.Y) + (-M.Z.Z * P.Z) + (-M.Z.W * P.W)
( M.W.X * P.X) + ( M.W.Y * P.Y) + ( M.W.Z * P.Z) + ( M.W.W * P.W)
```

Again filling in the known constants:
```
( M.X.X * ( 1,  0,  0,  0)) + ( M.X.Y * ( 0,  1,  0,  0)) + ( M.X.Z * ( 0,  0, -1,  0)) + ( M.X.W * ( 0,  0,  0,  1))
( M.Y.X * ( 1,  0,  0,  0)) + ( M.Y.Y * ( 0,  1,  0,  0)) + ( M.Y.Z * ( 0,  0, -1,  0)) + ( M.Y.W * ( 0,  0,  0,  1))
(-M.Z.X * ( 1,  0,  0,  0)) + (-M.Z.Y * ( 0,  1,  0,  0)) + (-M.Z.Z * ( 0,  0, -1,  0)) + (-M.Z.W * ( 0,  0,  0,  1))
( M.W.X * ( 1,  0,  0,  0)) + ( M.W.Y * ( 0,  1,  0,  0)) + ( M.W.Z * ( 0,  0, -1,  0)) + ( M.W.W * ( 0,  0,  0,  1))
```

This then simplifies down to:
```
( M.X.X * ( 1,  0,  0,  0)) + ( M.X.Y * ( 0,  1,  0,  0)) + ( M.X.Z * ( 0,  0, -1,  0)) + ( M.X.W * ( 0,  0,  0,  1))
( M.Y.X * ( 1,  0,  0,  0)) + ( M.Y.Y * ( 0,  1,  0,  0)) + ( M.Y.Z * ( 0,  0, -1,  0)) + ( M.Y.W * ( 0,  0,  0,  1))
(-M.Z.X * ( 1,  0,  0,  0)) + (-M.Z.Y * ( 0,  1,  0,  0)) + (-M.Z.Z * ( 0,  0, -1,  0)) + (-M.Z.W * ( 0,  0,  0,  1))
( M.W.X * ( 1,  0,  0,  0)) + ( M.W.Y * ( 0,  1,  0,  0)) + ( M.W.Z * ( 0,  0, -1,  0)) + ( M.W.W * ( 0,  0,  0,  1))
```

Since each step for each row is just `scalar * vector` it functionally means we simplify such that we have a matrix
```
 M.X.X  M.X.Y -M.X.Z  M.X.W
 M.Y.X  M.Y.Y -M.Y.Z  M.Y.W
-M.Z.X -M.Z.Y  M.Z.Z -M.Z.W
 M.W.X  M.W.Y -M.W.Z  M.W.W
```

Which is to say that the `Z` row/column, minus Z.Z, has been negated. All other elements were preserved as is.

--------------------------------

A helper can functionally simplify this to a single transform set where instead of the standard multiplication:
```csharp
public static Matrix4x4.Impl operator *(in Matrix4x4.Impl left, in Matrix4x4.Impl right)
{
    Matrix4x4.Impl result;

    result.X = Vector4.Transform(left.X, in right);
    result.Y = Vector4.Transform(left.Y, in right);
    result.Z = Vector4.Transform(left.Z, in right);
    result.W = Vector4.Transform(left.W, in right);

    return result;
}
```

We instead functionally have:
```csharp
public static Matrix4x4.Impl InverseHandedness(in Matrix4x4.Impl M)
{
    Matrix4x4.Impl P = new Matrix4x4.Impl(
        1, 0,  0, 0,
        0, 1,  0, 0,
        0, 0, -1, 0,
        0, 0,  0, 1
    );

    Matrix4x4.Impl result;

    result.X = Vector4.Transform( M.X, in P);
    result.Y = Vector4.Transform( M.Y, in P);
    result.Z = Vector4.Transform(-M.Z, in P);
    result.W = Vector4.Transform( M.W, in P);

    return result;
}
```

Which itself can be made even cheaper into the following, giving us simply 4 multiplications:
```csharp
public static Matrix4x4.Impl InverseHandedness(in Matrix4x4.Impl M)
{
    Matrix4x4.Impl result;

    result.X = M.X * Vector4.Create( 1,  1, -1,  1);
    result.Y = M.Y * Vector4.Create( 1,  1, -1,  1);
    result.Z = M.Z * Vector4.Create(-1, -1,  1, -1);
    result.W = M.W * Vector4.Create( 1,  1, -1,  1);

    return result;
}
```

Reply:
Yep, I agree with the optimization. I thought the explicit multiplies made it more clear but I don't mind changing that.

Apart from that change, these are the 4 scenarios you mentioned:

> we should be checking the raw results for Matrix4x4.*

Lets say we have some `args : Vector4` be the single argument (just to make things simple) to a `CreateRH` method. Then this is just `expected1 == CreateRH(args)`. This is the "base test".

> Matrix4x4.* transformed to a left-handed coordinate system is identical to Matrix4x4.*LeftHanded

This is `InvertHandedness(CreateRH(args)) == CreateLH(InvertHandednessVec(args))` (where <code>InvertHandedness(M) := P<sup>-1</sup> * M * P</code>, or the optimized version). By moving the coordinate inversion to the RHS and replacing the `CreateRH(args)` with `expected1`, we get `expected1 == InvertHandedness(CreateLH(InvertHandednessVec(args)))` which is the equivalent version from the PR. I preferred this rewrite just because it isolates the test to `CreateLH`, and the "base test" above already validates the substitution of `expected1` for `CreateRH` separately.

> and vice versa

So I think the suggestion here is to also validate `expected2 == CreateLH(args)`? That is, interpret `args` as an LH vector and do the same 2 tests as above. I avoided this because the coverage by doing just the first 2 tests was high (all the edge cases from the implementation were covered since they map 1-to-1) and I'd have to calculate the extra `expected2` matrix in every case :). It's not too hard to add though.